pkg/scheduler/framework/framework.go

```diff
@@ -143,8 +174,738 @@ func (f *framework) EventRecorder() record.EventRecorder {
+		if err := f.client.Update(ctx, binding, &client.UpdateOptions{}); err != nil {
+			return controller.NewAPIServerError(fmt.Errorf(errorFormat, binding.Name, err))
+		}
+	}
+	return nil
+}
+
+// markAsDeletingFor marks a list of bindings as deleting.
+func (f *framework) markAsDeletingFor(ctx context.Context, bindings []*fleetv1beta1.ClusterResourceBinding) error {
+	errorFormat := "failed to mark binding %s as deleting: %w"
+
+	for _, binding := range bindings {
+		// Deleting is a terminal state for a binding; since there is no point of return, and the
+		// scheduler has acknowledged its fate at this moment, it is safe for the scheduler to remove
+		// the scheduler cleanup finalizer from any binding that is of the deleting state.
+		controllerutil.RemoveFinalizer(binding, fleetv1beta1.SchedulerCleanupFinalizer)
+		binding.Spec.State = fleetv1beta1.BindingStateDeleting
+		if err := f.client.Update(ctx, binding, &client.UpdateOptions{}); err != nil {
+			return controller.NewAPIServerError(fmt.Errorf(errorFormat, binding.Name, err))
+		}
+	}
+	return nil
+}
+
+// sortByCreationTimestampBindings is a wrapper which implements Sort.Interface, which allows easy
+// sorting of bindings by their CreationTimestamps.
+type sortByCreationTimestampBindings []*fleetv1beta1.ClusterResourceBinding
+
+// Len() is for implementing Sort.Interface.
+func (s sortByCreationTimestampBindings) Len() int {
+	return len(s)
+}
+
+// Swap() is for implementing Sort.Interface.
+func (s sortByCreationTimestampBindings) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// Less() is for implementing Sort.Interface.
+func (s sortByCreationTimestampBindings) Less(i, j int) bool {
+	return s[i].CreationTimestamp.Before(&(s[j].CreationTimestamp))
+}
+
+// downscale performs downscaling on creating and active bindings, i.e., marks some of them as deleting.
+//
+// To minimize interruptions, the scheduler picks creating bindings first (in any order); if there are still more
+// bindings to trim, the scheduler prefers ones with smaller CreationTimestamps.
+func (f *framework) downscale(ctx context.Context, active, creating []*fleetv1beta1.ClusterResourceBinding, count int) (updatedActive, updatedCreating []*fleetv1beta1.ClusterResourceBinding, err error) {
+	if count == 0 {
+		// Skip if the downscale count is zero.
+		return active, creating, nil
+	}
+
+	// Trim creating bindings.
+	bindingsToDelete := make([]*fleetv1beta1.ClusterResourceBinding, 0, count)
+	for i := 0; i < len(creating) && i < count; i++ {
+		binding := creating[i]
+		bindingsToDelete = append(bindingsToDelete, binding)
+	}
+
+	if len(bindingsToDelete) == count {
+		// Trimming creating bindings alone would suffice.
+		return active, creating[count:], f.markAsDeletingFor(ctx, bindingsToDelete)
+	}
+
+	// Trimming obsolete and creating bindings alone is not enough, move on to active bindings.
+
+	// Sort the bindings by their CreationTimestamps.
+	sorted := sortByCreationTimestampBindings(active)
```

Comment:
I am not sure that the create time of the binding is the best indicator. What about last update time? Need to think through what's the real life implication of picking one bind vs another.

Reply:
Hi Ryan! The objectmeta does not have a lastUpdateTimestamp, there are only Creation and Deletion timestamps I think. We could use lastTransitionTime, but it is in the condition (which is manipulated by the Dispatcher). The alternative is to add a label of our own to mark lastUpdateTimestamp.